Feat/questions optional

rahat1994 opened the pull request:
issue #630.

"Currently, the questions feature is not optional, but some people, like me, may not want to use it, since Pinkary's initial proposal - I think, was for bio links."

rahat1994 commented:
@MrPunyapal do you think I need to do anything else? If so please let me know.

MrPunyapal commented:
Tests are missing
 
And also a validation in a questions/create component

xiCO2k commented:
I would call it `can_receive_questions`.

devajmeireles commented on tests/Unit/Models/UserTest.php:
```diff
@@ -43,6 +44,14 @@
     expect($user->is_verified)->toBeTrue();
 });
 
+test('prefers questions', function () {
```
Doesn't make sense. You're creating a property like true and asserting that true is true. Shouldn't we worry about writing a test validating the update action?

rahat1994 replied:
fixed

devajmeireles commented on app/Http/Controllers/UserController.php:
```diff
@@ -42,8 +42,10 @@ public function show(User $user): View
+        $data = $request->validated();
+        $data['prefers_questions'] = $request->has('prefers_questions');
 
-        $user->fill($request->validated());
```
Since the `prefers_questions` is validated as a boolean. Why not preserve the original code?

rahat1994 replied:
Because if the checkbox is unchecked the form doesn't include it in the payload. So I check if the prefers_questions field is set it will store the value, If not it will store 0 or false.

MrPunyapal replied:
You can use radio or select if you want to preserve previous code

rahat1994 replied:
Okay, I think I will use select.

devajmeireles commented on app/Http/Requests/UserUpdateRequest.php:
```diff
@@ -29,16 +29,26 @@ public function rules(): array
             'mail_preference_time' => [Rule::enum(UserMailPreference::class)],
             'bio' => ['nullable', 'string', 'max:255'],
             'prefers_anonymous_questions' => ['required', 'boolean'],
+            'prefers_questions' => ['nullable', 'boolean'],
```
Your migration doesn't set the `prefers_questions` as nullable, so why should this be nullable?

rahat1994 replied:
will update the line. Thanks for pointing out.

MrPunyapal commented on resources/views/profile/show.blade.php:
```diff
@@ -2,7 +2,9 @@
             <livewire:links.index :userId="$user->id" />
-            <livewire:questions.create :toId="$user->id" />
+            @if($user->prefers_questions)
+                <livewire:questions.create :toId="$user->id" />
```
Make sure the user can share an update even if the user is not wanting to receive questions.

rahat1994 replied:
I am doing an user check if the user is the same as logged in user it will show the form.

MrPunyapal commented on resources/views/profile/show.blade.php:
```diff
@@ -2,7 +2,9 @@
     <div class="flex flex-col items-center justify-center py-10">
         <div class="min-h-screen w-full max-w-md overflow-hidden rounded-lg px-4 dark:shadow-md md:px-0">
             <livewire:links.index :userId="$user->id" />
-            <livewire:questions.create :toId="$user->id" />
```
Also make some related changes in question/create component.

For example if I have opened a profile of the user's when he wants to receive questions. Before writing the question and clicking on the send button if that user disables the option I should get a respected message.

rahat1994 replied:
I have added required validations to question/create.

But for your example I have to poll the database at regular intervals would you be okay with that?

thanks again.

MrPunyapal replied:
Why you wanna poll?

When you just check when someone clicks on the send button?

rahat1994 commented:
Things I changed

1. Added a select for choosing `prefers question`.
2. Added necessary validation rules.
3. Even if they don't want to receive questions when they visit their own profile they can share an update.
4. before storing a question It is checked if the user accepts question.
5. Added tests to see if the users accepts test.

MrPunyapal commented on resources/views/profile/partials/update-profile-information-form.blade.php:
```diff
@@ -145,10 +145,28 @@ class="mt-2"
             <x-input-label
                 for="prefers_anonymous_questions"
-                :value="__('How would you like to do questions by default?')"
+                :value="__('How would you like to do questions by default? Will only work when you accept questions.')"
```
why?

MrPunyapal replied:
prefers_anonymoud_question applicable when you ask questions.

So both are not related right now!

rahat1994 replied:
Oh okay, I thought they are related. Thanks for clarifying. I will reverse that line.

MrPunyapal commented on resources/views/profile/partials/update-profile-information-form.blade.php:
```diff
@@ -145,10 +145,28 @@ class="mt-2"
+                name="prefers_questions"
+                class="mt-1 block w-full"
+                class="mt-1 block w-full"
+                :options="[true => __('People can ask me questions'), false => __('I do not want to receive questions')]"
```
keep it simple "yes" and "no"

MrPunyapal commented on app/Livewire/Questions/Create.php:
```diff
@@ -221,7 +227,7 @@ public function store(Request $request): void
         $validated = $this->validate([
             'anonymously' => ['boolean', Rule::excludeIf($this->isSharingUpdate)],
-            'content' => ['required', 'string', 'min: 3', 'max:'.$this->maxContentLength, new NoBlankCharacters],
+            'content' => ['required', 'string', 'min: 3', 'max:' . $this->maxContentLength, new NoBlankCharacters],
```
undo

MrPunyapal commented on app/Http/Requests/UserUpdateRequest.php:
```diff
@@ -29,15 +29,25 @@ public function rules(): array
+                'min:4',
+                'max:50',
+                Rule::unique(User::class)->ignore($user->id),
                 new Username($user),
```
undo

MrPunyapal commented on app/Http/Requests/UserUpdateRequest.php:
```diff
@@ -29,15 +29,25 @@ public function rules(): array
             'username' => [
-                'required', 'string', 'min:4', 'max:50', Rule::unique(User::class)->ignore($user->id),
+                'required',
+                'string',
+                'min:4',
+                'max:50',
+                Rule::unique(User::class)->ignore($user->id),
                 new Username($user),
             ],
             'email' => [
-                'required', 'string', 'lowercase', 'email', 'max:255', Rule::unique(User::class)->ignore($user->id),
+                'required',
+                'string',
+                'lowercase',
+                'email',
+                'max:255',
+                Rule::unique(User::class)->ignore($user->id),
```
undo

MrPunyapal replied:
undo this

MrPunyapal commented on app/Livewire/Questions/Create.php:
```diff
@@ -206,6 +205,13 @@ public function store(Request $request): void
 
+        $toUser = User::findOrFail($this->toId);
+
+        if (!$toUser->prefers_questions) {
```
please run Pint
